### PR TITLE
[formatter] made mechanism for finding text regions more powerful

### PR DESCRIPTION
**Pull request by meysholdt**
Signed-off-by: Moritz Eysholdt <moritz.eysholdt@itemis.de>

### CODE REVIEW

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/AbstractSemanticRegionsFinder.java [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+		Set<Predicate<ISemanticRegion>> predicates = Sets.newHashSet();
+		for (int i = 0; i < features.length; i++) {
+			EStructuralFeature feat = features[i];
+			assertNoContainment(feat);
+			predicates.add(new FeaturePredicate(feat));
+		}
+		return findAll(Predicates.or(predicates));
+	}
+
+	protected ImmutableList<ISemanticRegion> findAll(Predicate<ISemanticRegion> predicate) {
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/AbstractSemanticRegionsFinder.java]: findAll will never return more than the first element, is that intentional?

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/ITextRegionExtensions.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.xtext.Action;
+import org.eclipse.xtext.ParserRule;
+import org.eclipse.xtext.RuleCall;
+
+/**
+ * @author Moritz Eysholdt - Initial contribution and API
+ */
+public interface ITextRegionExtensions {
```
Comment (szarnekow) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/ITextRegionExtensions.java]: Why was this moved to a new interface? ITextRegionAccess now has an accessor to obtain the extensions but in fact the methods here would be easier to discover if they were still on the *access interface - after all, they provide access to regions by means of finders.

Comment (meysholdt) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/ITextRegionExtensions.java]: Discussed offline to keep it this way; Methods from the *Extensions interface are intended to be on Xtend's extension scope while most methods from *Access should be on the extensions scope and some not even on the implicit scope.

File: plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/AbstractSemanticRegionsFinder.java (hunk 1 above)
Comment (meysholdt) [inline plugins/org.eclipse.xtext/src/org/eclipse/xtext/formatting2/regionaccess/internal/AbstractSemanticRegionsFinder.java]: refactored this in 1d942e3 to be an abstract method. Now subclasses get to decide on an implementation explicitly.

**Comment by meysholdt** [comment]
Jenkins test
